helix-core/src/main/java/org/apache/helix/controller/stages/BestPossibleStateCalcStage.java

```diff
@@ -249,8 +253,15 @@ private boolean validateOfflineInstancesLimit(final ResourceControllerDataProvid
     // Init rebalancer with the rebalance preferences.
     Map<ClusterConfig.GlobalRebalancePreferenceKey, Integer> preferences = cache.getClusterConfig()
         .getGlobalRebalancePreference();
+
+    // TODO: Add an appropriate entityName
+    if (METRIC_COLLECTOR_THREAD_LOCAL.get() == null) {
+      METRIC_COLLECTOR_THREAD_LOCAL.set(new MetricCollector(helixManager.getClusterName(), ""));
```

Comment:
I guess the lazy init is not required here.
Let's just put it to the constructor of the stage object.

It's always better to have the metrics code "disappear", you know.